File: app/services/grok/tool_call.py
"""
Tool call utilities for OpenAI-compatible function calling.

Provides prompt-based emulation of tool calls by injecting tool definitions
into the system prompt and parsing structured responses.
"""

import json
import re
import uuid
from typing import Any, Dict, List, Optional, Tuple


def build_tool_prompt(
    tools: List[Dict[str, Any]],
    tool_choice: Optional[Any] = None,
    parallel_tool_calls: bool = True,
) -> str:
    """Generate a system prompt block describing available tools.

    Args:
        tools: List of OpenAI-format tool definitions.
        tool_choice: "auto", "required", "none", or {"type":"function","function":{"name":"..."}}.
        parallel_tool_calls: Whether multiple tool calls are allowed.

    Returns:
        System prompt string to prepend to the conversation.
    """
    if not tools:
        return ""

    if tool_choice == "none":
        return ""

    lines = [
        "# Available Tools",
        "",
        "You have access to the following tools. To call a tool, output a <tool_call> block with a JSON object containing \"name\" and \"arguments\".",
        "",
        "Format:",
        "<tool_call>",
        '{"name": "function_name", "arguments": {"param": "value"}}',
        "</tool_call>",
        "",
    ]

    if parallel_tool_calls:
        lines.append("You may make multiple tool calls in a single response by using multiple <tool_call> blocks.")
        lines.append("")

    lines.append("## Tool Definitions")
    lines.append("")
    for tool in tools:
        if tool.get("type") != "function":
            continue
        func = tool.get("function", {})
        name = func.get("name", "")
        desc = func.get("description", "")
        params = func.get("parameters", {})

        lines.append(f"### {name}")
        if desc:
            lines.append(f"{desc}")
        if params:
            lines.append(f"Parameters: {json.dumps(params, ensure_ascii=False)}")
        lines.append("")

    if tool_choice == "required":
        lines.append("IMPORTANT: You MUST call at least one tool in your response. Do not respond with only text.")
    elif isinstance(tool_choice, dict):
        func_info = tool_choice.get("function", {})
        forced_name = func_info.get("name", "")
        if forced_name:
            lines.append(f"IMPORTANT: You MUST call the tool \"{forced_name}\" in your response.")
    else:
        lines.append("Decide whether to call a tool based on the user's request. If you don't need a tool, respond normally with text only.")

    lines.append("")
    lines.append("When you call a tool, you may include text before or after the <tool_call> blocks, but the tool call blocks must be valid JSON.")

    return "\n".join(lines)


_TOOL_CALL_RE = re.compile(
    r"<tool_call>\s*(.*?)\s*</tool_call>",
    re.DOTALL,
)


def _strip_code_fences(text: str) -> str:
    if not text:
        return text
    cleaned = text.strip()
    if cleaned.startswith("```"):
        cleaned = re.sub(r"^```[a-zA-Z0-9_-]*\s*", "", cleaned)
        cleaned = re.sub(r"\s*```$", "", cleaned)
    return cleaned.strip()


def _extract_json_object(text: str) -> str:
    if not text:
        return text
    start = text.find("{")
    if start == -1:
        return text
    end = text.rfind("}")
    if end == -1:
        return text[start:]
    if end < start:
        return text
    return text[start : end + 1]


def _remove_trailing_commas(text: str) -> str:
    if not text:
        return text
    return re.sub(r",\s*([}\]])", r"\1", text)


def _balance_braces(text: str) -> str:
    if not text:
        return text
    open_count = 0
    close_count = 0
    in_string = False
    escape = False
    for ch in text:
        if escape:
            escape = False
            continue
        if ch == "\\" and in_string:
            escape = True
            continue
        if ch == '"':
            in_string = not in_string
            continue
        if in_string:
            continue
        if ch == "{":
            open_count += 1
        elif ch == "}":
            close_count += 1
    if open_count > close_count:
        text = text + ("}" * (open_count - close_count))
    return text


def _repair_json(text: str) -> Optional[Any]:
    if not text:
        return None
    cleaned = _strip_code_fences(text)
    cleaned = _extract_json_object(cleaned)
    cleaned = cleaned.replace("\r\n", "\n").replace("\r", "\n")
    cleaned = cleaned.replace("\n", " ")
    cleaned = _remove_trailing_commas(cleaned)
    cleaned = _balance_braces(cleaned)
    try:
        return json.loads(cleaned)
    except json.JSONDecodeError:
        return None


def parse_tool_call_block(
    raw_json: str,
    tools: Optional[List[Dict[str, Any]]] = None,
) -> Optional[Dict[str, Any]]:
    """Parse a single tool call JSON block."""
    if not raw_json:
        return None
    parsed = None
    try:
        parsed = json.loads(raw_json)
    except json.JSONDecodeError:
        parsed = _repair_json(raw_json)
    if not isinstance(parsed, dict):
        return None

    name = parsed.get("name")
    arguments = parsed.get("arguments", {})
    if not name:
        return None

    valid_names = set()
    if tools:
        for tool in tools:
            func = tool.get("function", {})
            tool_name = func.get("name")
            if tool_name:
                valid_names.add(tool_name)
    if valid_names and name not in valid_names:
        return None

    if isinstance(arguments, dict):
        arguments_str = json.dumps(arguments, ensure_ascii=False)
    elif isinstance(arguments, str):
        arguments_str = arguments
    else:
        arguments_str = json.dumps(arguments, ensure_ascii=False)

    return {
        "id": f"call_{uuid.uuid4().hex[:24]}",
        "type": "function",
        "function": {"name": name, "arguments": arguments_str},
    }


def parse_tool_calls(
    content: str,
    tools: Optional[List[Dict[str, Any]]] = None,
) -> Tuple[Optional[str], Optional[List[Dict[str, Any]]]]:
    """Parse tool call blocks from model output.

    Returns:
        Tuple of (text_content, tool_calls_list).
    """
    if not content:
        return content, None

    matches = list(_TOOL_CALL_RE.finditer(content))
    if not matches:
        return content, None

    tool_calls = []
    for match in matches:
        raw_json = match.group(1).strip()
        tool_call = parse_tool_call_block(raw_json, tools)
        if tool_call:
            tool_calls.append(tool_call)

    if not tool_calls:
        return content, None

    text_parts = []
    last_end = 0
    for match in matches:
        before = content[last_end:match.start()]
        if before.strip():
            text_parts.append(before.strip())
        last_end = match.end()
    trailing = content[last_end:]
    if trailing.strip():
        text_parts.append(trailing.strip())

    text_content = "\n".join(text_parts) if text_parts else None

    return text_content, tool_calls


def build_tool_overrides(tools: List[Dict[str, Any]]) -> Dict[str, Any]:
    """Convert OpenAI tool format to Grok's toolOverrides format (experimental)."""
    if not tools:
        return {}

    tool_overrides = {}
    for tool in tools:
        if tool.get("type") != "function":
            continue
        func = tool.get("function", {})
        name = func.get("name", "")
        if not name:
            continue
        tool_overrides[name] = {
            "enabled": True,
            "description": func.get("description", ""),
            "parameters": func.get("parameters", {}),
        }

    return tool_overrides


def format_tool_history(messages: List[Dict[str, Any]]) -> List[Dict[str, Any]]:
    """Convert assistant messages with tool_calls and tool role messages into text format.

    Since Grok's web API only accepts a single message string, this converts
    tool-related messages back to a text representation for multi-turn conversations.
    """
    result = []
    for msg in messages:
        role = msg.get("role", "")
        content = msg.get("content")
        tool_calls = msg.get("tool_calls")
        tool_call_id = msg.get("tool_call_id")
        name = msg.get("name")

        if role == "assistant" and tool_calls:
            parts = []
            if content:
                parts.append(content if isinstance(content, str) else str(content))
            for tc in tool_calls:
                func = tc.get("function", {})
                tc_name = func.get("name", "")
                tc_args = func.get("arguments", "{}")
                parts.append(f'<tool_call>{{"name":"{tc_name}","arguments":{tc_args}}}</tool_call>')
            result.append({
                "role": "assistant",
                "content": "\n".join(parts),
            })

        elif role == "tool":
            tool_name = name or "unknown"
            call_id = tool_call_id or ""
            content_str = content if isinstance(content, str) else json.dumps(content, ensure_ascii=False) if content else ""
            result.append({
                "role": "user",
                "content": f"tool ({tool_name}, {call_id}): {content_str}",
            })

        else:
            result.append(msg)

    return result


__all__ = [
    "build_tool_prompt",
    "parse_tool_calls",
    "parse_tool_call_block",
    "build_tool_overrides",
    "format_tool_history",
]

File: src/grok/toolCall.ts
/**
 * Tool call utilities for OpenAI-compatible function calling.
 *
 * Provides prompt-based emulation of tool calls by injecting tool definitions
 * into the system prompt and parsing structured responses.
 */

export interface ToolDefinition {
  type?: string;
  function?: {
    name?: string;
    description?: string;
    parameters?: Record<string, unknown>;
  };
}

export interface ToolCall {
  id: string;
  type: "function";
  function: { name: string; arguments: string };
}

export interface OpenAIMessage {
  role: string;
  content?: string | unknown;
  tool_calls?: Array<{
    function?: { name?: string; arguments?: string };
  }>;
  tool_call_id?: string;
  name?: string;
}

const TOOL_CALL_RE = /<tool_call>\s*([\s\S]*?)\s*<\/tool_call>/g;

function generateCallId(): string {
  const hex = crypto.randomUUID().replace(/-/g, "").slice(0, 24);
  return `call_${hex}`;
}

export function buildToolPrompt(
  tools: ToolDefinition[],
  toolChoice?: unknown,
  parallelToolCalls: boolean = true,
): string {
  if (!tools || !tools.length) return "";
  if (toolChoice === "none") return "";

  const lines: string[] = [
    "# Available Tools",
    "",
    'You have access to the following tools. To call a tool, output a <tool_call> block with a JSON object containing "name" and "arguments".',
    "",
    "Format:",
    "<tool_call>",
    '{"name": "function_name", "arguments": {"param": "value"}}',
    "</tool_call>",
    "",
  ];

  if (parallelToolCalls) {
    lines.push("You may make multiple tool calls in a single response by using multiple <tool_call> blocks.");
    lines.push("");
  }

  lines.push("## Tool Definitions");
  lines.push("");
  for (const tool of tools) {
    if (tool.type !== "function") continue;
    const func = tool.function ?? {};
    const name = func.name ?? "";
    const desc = func.description ?? "";
    const params = func.parameters;

    lines.push(`### ${name}`);
    if (desc) lines.push(desc);
    if (params) lines.push(`Parameters: ${JSON.stringify(params)}`);
    lines.push("");
  }

  if (toolChoice === "required") {
    lines.push("IMPORTANT: You MUST call at least one tool in your response. Do not respond with only text.");
  } else if (typeof toolChoice === "object" && toolChoice !== null) {
    const funcInfo = (toolChoice as Record<string, any>).function ?? {};
    const forcedName = funcInfo.name ?? "";
    if (forcedName) {
      lines.push(`IMPORTANT: You MUST call the tool "${forcedName}" in your response.`);
    }
  } else {
    lines.push("Decide whether to call a tool based on the user's request. If you don't need a tool, respond normally with text only.");
  }

  lines.push("");
  lines.push("When you call a tool, you may include text before or after the <tool_call> blocks, but the tool call blocks must be valid JSON.");

  return lines.join("\n");
}

// --- JSON repair utilities ---

function stripCodeFences(text: string): string {
  if (!text) return text;
  let cleaned = text.trim();
  if (cleaned.startsWith("```")) {
    cleaned = cleaned.replace(/^```[a-zA-Z0-9_-]*\s*/, "");
    cleaned = cleaned.replace(/\s*```$/, "");
  }
  return cleaned.trim();
}

function extractJsonObject(text: string): string {
  if (!text) return text;
  const start = text.indexOf("{");
  if (start === -1) return text;
  const end = text.lastIndexOf("}");
  if (end === -1) return text.slice(start);
  if (end < start) return text;
  return text.slice(start, end + 1);
}

function removeTrailingCommas(text: string): string {
  if (!text) return text;
  return text.replace(/,\s*([}\]])/g, "$1");
}

function balanceBraces(text: string): string {
  if (!text) return text;
  let openCount = 0;
  let closeCount = 0;
  let inString = false;
  let escape = false;
  for (const ch of text) {
    if (escape) { escape = false; continue; }
    if (ch === "\\" && inString) { escape = true; continue; }
    if (ch === '"') { inString = !inString; continue; }
    if (inString) continue;
    if (ch === "{") openCount++;
    else if (ch === "}") closeCount++;
  }
  if (openCount > closeCount) {
    return text + "}".repeat(openCount - closeCount);
  }
  return text;
}

function repairJson(text: string): unknown | null {
  if (!text) return null;
  let cleaned = stripCodeFences(text);
  cleaned = extractJsonObject(cleaned);
  cleaned = cleaned.replace(/\r\n/g, "\n").replace(/\r/g, "\n").replace(/\n/g, " ");
  cleaned = removeTrailingCommas(cleaned);
  cleaned = balanceBraces(cleaned);
  try {
    return JSON.parse(cleaned);
  } catch {
    return null;
  }
}

// --- Parsing ---

export function parseToolCallBlock(
  rawJson: string,
  tools?: ToolDefinition[],
): ToolCall | null {
  if (!rawJson) return null;

  let parsed: unknown;
  try {
    parsed = JSON.parse(rawJson);
  } catch {
    parsed = repairJson(rawJson);
  }
  if (!parsed || typeof parsed !== "object") return null;

  const obj = parsed as Record<string, unknown>;
  const name = obj.name;
  if (typeof name !== "string" || !name) return null;

  // Validate against known tool names
  if (tools && tools.length) {
    const validNames = new Set<string>();
    for (const tool of tools) {
      const tn = tool.function?.name;
      if (tn) validNames.add(tn);
    }
    if (validNames.size && !validNames.has(name)) return null;
  }

  const args = obj.arguments ?? {};
  let argumentsStr: string;
  if (typeof args === "string") {
    argumentsStr = args;
  } else {
    argumentsStr = JSON.stringify(args);
  }

  return {
    id: generateCallId(),
    type: "function",
    function: { name, arguments: argumentsStr },
  };
}

export function parseToolCalls(
  content: string,
  tools?: ToolDefinition[],
): { text: string | null; toolCalls: ToolCall[] | null } {
  if (!content) return { text: content, toolCalls: null };

  const matches: Array<{ start: number; end: number; json: string }> = [];
  const re = new RegExp(TOOL_CALL_RE.source, TOOL_CALL_RE.flags);
  let m: RegExpExecArray | null;
  while ((m = re.exec(content)) !== null) {
    matches.push({ start: m.index, end: m.index + m[0].length, json: (m[1] ?? "").trim() });
  }

  if (!matches.length) return { text: content, toolCalls: null };

  const toolCalls: ToolCall[] = [];
  for (const match of matches) {
    const tc = parseToolCallBlock(match.json, tools);
    if (tc) toolCalls.push(tc);
  }

  if (!toolCalls.length) return { text: content, toolCalls: null };

  // Extract text outside of tool_call blocks
  const textParts: string[] = [];
  let lastEnd = 0;
  for (const match of matches) {
    const before = content.slice(lastEnd, match.start);
    if (before.trim()) textParts.push(before.trim());
    lastEnd = match.end;
  }
  const trailing = content.slice(lastEnd);
  if (trailing.trim()) textParts.push(trailing.trim());

  const textContent = textParts.length ? textParts.join("\n") : null;
  return { text: textContent, toolCalls };
}

export function formatToolHistory(messages: OpenAIMessage[]): OpenAIMessage[] {
  const result: OpenAIMessage[] = [];
  for (const msg of messages) {
    const role = msg.role ?? "";
    const content = msg.content;
    const toolCalls = msg.tool_calls;
    const toolCallId = msg.tool_call_id;
    const name = msg.name;

    if (role === "assistant" && toolCalls && toolCalls.length) {
      const parts: string[] = [];
      if (content) {
        parts.push(typeof content === "string" ? content : String(content));
      }
      for (const tc of toolCalls) {
        const func = tc.function ?? {};
        const tcName = func.name ?? "";
        const tcArgs = func.arguments ?? "{}";
        parts.push(`<tool_call>{"name":"${tcName}","arguments":${tcArgs}}</tool_call>`);
      }
      result.push({ role: "assistant", content: parts.join("\n") });
    } else if (role === "tool") {
      const toolName = name || "unknown";
      const callId = toolCallId || "";
      const contentStr = typeof content === "string" ? content : content ? JSON.stringify(content) : "";
      result.push({ role: "user", content: `tool (${toolName}, ${callId}): ${contentStr}` });
    } else {
      result.push(msg);
    }
  }
  return result;
}
